### ISSUE TITLE
Assertion `ticks_now == trace_ticks' failed to hold

### ISSUE THREAD

**Issue by hotsphink** [mozilla/rr#2186]
I am getting this error when reverse-continuing a patched spidermonkey shell (a bit of a weird one; it's the one whose JIT generates ARM code that it then runs under a simulator. But that's fine; it's my locally patched one that is exhibiting this behavior.):

[FATAL /home/sfink/src/rr/src/ReplaySession.cc:1023:check_ticks_consistency()]
 (task 6105 (rec:5866) at time 550)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'SYSCALL: write'; expected 109121266, got 109121268

Here are steps to reproduce, though I still don't have a good way to host files so I'm trying out Firefox Send for the first time, which means that the first person who downloads this gets it (I can re-send it somehow as needed):

    have a gecko checkout
    cd js/src/jit-test
    wget -Ojs https://send.firefox.com/download/a9ca0c4681/#eFSkcoDc8aJnRoQpZxQQDQ
    chmod +x ./js
    rr record ./js --ion-eager --ion-offthread-compile=off tests/ion/lookupswitch.js
    rr replay
    c
    # Go to crash
    f 8
    watch -l *(void**)&head_
    rc # still at seg fault
    rc

Here's the dump around that tick count (looks like it didn't make it very far before the crash):

    {
      real_time:964798.918677 global_time:543, event:`SYSCALLBUF_RESET' tid:6357, ticks:67218238
    }
    {
      real_time:964798.918746 global_time:544, event:`SYSCALL: mmap2' (state:EXITING_SYSCALL) tid:6357, ticks:67218238
    eax:0x4983d000 ecx:0x10000 edx:0x5 ebx:0x4983d000 esp:0x681ffe90 ebp:0x0 esi:0x32 edi:0xffffffff eip:0x70000002 eflags:0x246 xcs:0x23 xss:0x2b xds:
    0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0xc0
    }
    {
      real_time:964798.975754 global_time:545, event:`SYSCALLBUF_FLUSH' tid:6357, ticks:94461720
    }
    {
      real_time:964798.975794 global_time:546, event:`SYSCALL: mmap2' (state:ENTERING_SYSCALL) tid:6357, ticks:94461720
    eax:0xffffffda ecx:0x10000 edx:0x5 ebx:0x4988d000 esp:0x681ffe90 ebp:0x0 esi:0x32 edi:0xffffffff eip:0x70000002 eflags:0x246 xcs:0x23 xss:0x2b xds:
    0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0xc0
    }
    {
      real_time:964798.975802 global_time:547, event:`SYSCALLBUF_RESET' tid:6357, ticks:94461720
    }
    {
      real_time:964798.975890 global_time:548, event:`SYSCALL: mmap2' (state:EXITING_SYSCALL) tid:6357, ticks:94461720
    eax:0x4988d000 ecx:0x10000 edx:0x5 ebx:0x4988d000 esp:0x681ffe90 ebp:0x0 esi:0x32 edi:0xffffffff eip:0x70000002 eflags:0x246 xcs:0x23 xss:0x2b xds:
    0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0xc0
    }
    {
      real_time:964799.002141 global_time:549, event:`SYSCALLBUF_FLUSH' tid:6357, ticks:109121346
    }
    {
      real_time:964799.002160 global_time:550, event:`SYSCALL: write' (state:ENTERING_SYSCALL) tid:6357, ticks:109121346
    eax:0xffffffda ecx:0xfff706bc edx:0x6c ebx:0x2 esp:0x681ffe90 ebp:0xfff706bc esi:0x0 edi:0xf7bb3ce0 eip:0x70000002 eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0x4
    }
    {
      real_time:964799.002167 global_time:551, event:`SYSCALLBUF_RESET' tid:6357, ticks:109121346
    }
    {
      real_time:964799.002240 global_time:552, event:`SYSCALL: write' (state:EXITING_SYSCALL) tid:6357, ticks:109121346
    eax:0x6c ecx:0xfff706bc edx:0x6c ebx:0x2 esp:0x681ffe90 ebp:0xfff706bc esi:0x0 edi:0xf7bb3ce0 eip:0x70000002 eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0x4
    }
    {
      real_time:964799.002368 global_time:553, event:`SIGNAL: SIGSEGV(det)' tid:6357, ticks:109121698
    eax:0x0 ecx:0xf7bb4884 edx:0x0 ebx:0x9c77ff4 esp:0xfff72c40 ebp:0xfff72c48 esi:0xd edi:0x10 eip:0x856f6b6 eflags:0x10286 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0xffffffff
    }
    {
      real_time:964799.002451 global_time:554, event:`SIGNAL_HANDLER: SIGSEGV(det)' tid:6357, ticks:109121698
    eax:0xb ecx:0xfff726cc edx:0xfff7264c ebx:0x9c77ff4 esp:0xfff7263c ebp:0xfff72c48 esi:0xd edi:0x10 eip:0x899c5c0 eflags:0x286 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0xffffffff st0:0x0 st1:0x0 st2:0x0 st3:0x0 st4:0x0 st5:0x0 st6:0x0 st7:0x0 ymm0:0x0 ymm1:0x0 ymm2:0x0 ymm3:0x0 ymm4:0x0 ymm5:0x0 ymm6:0x0 ymm7:0x0
    }

**Comment by rocallahan** [comment]
If I set a conditional breakpoint that stops at the third memset, then the bug goes away. However, that could just be the frequent conditional breakpoint stops interfering with things.

**Comment by rocallahan** [comment]
Needless to say, running `rr replay -a` is fine.

**Comment by rocallahan** [comment]
I see a lot of PerfCounter warnings like `Measured too many ticks; measure=1412675, interrupt=1412645` in the logs. But most of them aren't causing problems.

**Comment by rocallahan** [comment]
I should mention that I can reproduce this problem just by advancing to the start of the execution leading up to the event where we diverge, setting the watchpoint, and then continuing. It's not specifically related to reverse execution, though it could be related to `fix_watchpoint_coalescing_quirk`.

**Comment by rocallahan** [comment]
And disabling `fix_watchpoint_coalescing_quirk` doesn't fix it.

**Comment by rocallahan** [comment]
I'm beginning to fear some kind of architecture bug here. When we run to the instruction where `js::jit::CodeGenerator::visitInterruptCheck` calls `js::LifoAlloc::allocInfallible` from the start of an event, with no watchpoints set we get N ticks but with a watchpoint set we get N + 1 ticks. If we do the same thing but run to where `js::LifoAlloc::allocInfallible` has just returned into `js::jit::CodeGenerator::visitInterruptCheck`, with no watchpoints we get M (> N) ticks and with the watchpoint set we get M + 2 ticks. But if I set a breakpoint anywhere in the vicinity of the code leading up to the stopping point, the bug goes away.

**Comment by rocallahan** [comment]
It's difficult to narrow the bug down further because `js::LifoAlloc::allocInfallible` gets called a lot and if I set a breakpoint inside it, even a conditional one, those stops make the bug go away.

**Comment by rocallahan** [comment]
We do hit the watchpoint a couple of times but that seems irrelevant since if I disable the watchpoint after the last time it is hit, the bug goes away.

**Comment by rocallahan** [comment]
Interestingly, setting a read watchpoint on the location triggers the same behavior, but setting a hardware breakpoint does not.

**Comment by rocallahan** [comment]
OK I'm pretty this is a hardware bug. The key observation is that the problematic call to `js::LifoAlloc::allocInfallible` performs a memset implemented as `rep stos` which writes to 36 bytes, and the end of this area is close to the address of our watchpoint. If I move the watchpoint address the number of overcounts changes like so:
```
0xf6036e90: 108564786 // the memset ends at this address
0xf6036e98: 108564786
0xf6036ea0: 108564787
0xf6036ea8: 108564787
0xf6036eb0: 108564787
0xf6036eb8: 108564787
0xf6036ec0: 108564786
0xf6036ec8: 108564786
0xf6036ed0: 108564786 // where the watchpoint is placed during the testcase in this bug
0xf6036ed8: 108564786
0xf6036ee0: 108564785
0xf6036ee8: 108564785
0xf6036ef0: 108564784 // these values are correct
0xf6036ef8: 108564784
0xf6036e00: 108564784
```

**Comment by rocallahan** [comment]
A watchpoint at `0xf6036e9c` produces a count of 108564787, so it looks like the threshold is actually 4-byte aligned, which is consistent with the start address of the memset being `0xf6036e6c` i.e. 4-byte aligned. Note that this testcase is in 32-bit mode.

**Comment by rocallahan** [comment]
The offset from the `rep movs` start to the first "safe" address is exactly 128 bytes.

**Comment by rocallahan** [comment]
I should probably try writing a standalone testcase to help narrow down the conditions under which this bug can arise. There probably isn't going to be much we can do about it in rr though. Maybe allowing some slop...

**Comment by hotsphink** [comment]
Uh, wow. That went deeper than I expected. Thanks; at least I know enough now to work around it if I run into it again.

**Comment by rocallahan** [comment]
What CPU do you have? This standalone program shows the bug on my Skylake laptop: https://gist.github.com/rocallahan/32561514df847ec3b54370fe7553500d
The bug does not seem to be present on a Broadwell machine I have access to.

**Comment by rocallahan** [comment]
Actually this looks like it's covered by a Skylake erratum: https://www.intel.com/content/dam/www/public/us/en/documents/specification-updates/desktop-6th-gen-core-family-spec-update.pdf
Which I blogged about two years ago! https://robert.ocallahan.org/2016/04/skylake-erratum-affecting-rr.html

**Comment by rocallahan** [comment]
Of course bugs being triggered by watchpoints is not documented in the erratum, so that is new information.

**Comment by rocallahan** [comment]
I blogged about this: https://robert.ocallahan.org/2018/05/intel-cpu-bug-affecting-rr-watchpoints.html

I guess we could try to work around this by detecting divergence and then automatically reverting to a previous checkpoint and replaying again with watchpoints disabled. That would probably let us detect that we're hitting this bug, but then what? We could die with a more informative message, but that isn't much help to a user who hits this.

It is possible for the kernel to disable the CPU's fast-string optimization, which would probably fix this. But I don't think there's a way to manually force that in Linux, currently, not even a command-line parameter. I guess the most expedient solution would to be create one, though I'm not particularly excited about doing that.

**Comment by rocallahan** [comment]
But I'd be delighted if someone else worked up the kernel patch and submitted it upstream!

**Comment by hotsphink** [comment]
Yes. I have a Skylake. /proc/cpuinfo says: Intel(R) Xeon(R) CPU E3-1505M v5 @ 2.80GHz

Thanks for tracking it down.

**Comment by hotsphink** [comment]
I wonder if this could be avoided with appropriate usage of -mstringop-strategy (or -mmemcpy-strategy + -mmemset-strategy).

**Comment by hotsphink** [comment]
"avoided" as in, a workaround for the program being debugged.

**Comment by rocallahan** [comment]
Yeah if you avoid using string instructions, shouldn't be a problem.

I guess rr could try using LD_PRELOAD substitution to override memcpy/memset during recording, but that's a bit scary...
